pkg/datapath/linux/devices_controller.go

```diff
@@ -0,0 +1,697 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+//go:build linux
+
+package linux
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/vishvananda/netlink"
+	"github.com/vishvananda/netns"
+	"golang.org/x/exp/maps"
+	"golang.org/x/exp/slices"
+	"golang.org/x/sys/unix"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/asm"
+
+	"github.com/cilium/cilium/pkg/backoff"
+	"github.com/cilium/cilium/pkg/datapath/linux/probes"
+	"github.com/cilium/cilium/pkg/datapath/tables"
+	"github.com/cilium/cilium/pkg/defaults"
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/hive/cell"
+	"github.com/cilium/cilium/pkg/ip"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/mac"
+	"github.com/cilium/cilium/pkg/statedb"
+)
+
+// DevicesControllerCell registers a controller that subscribes to network devices
+// and routes via netlink and populates the devices and routes devices.
+var DevicesControllerCell = cell.Module(
+	"datapath-devices-controller",
+	"Populates the device and route tables from netlink",
+
+	cell.Provide(
+		newDevicesController,
+		newDeviceManager,
+	),
+	cell.Invoke(func(*devicesController) {}),
+)
+
+var (
+	// batchingDuration is the amount of time to wait for more
+	// addr/route/link updates before processing the batch.
+	batchingDuration = 100 * time.Millisecond
+
+	// netlinkBuffer is the number of updates to buffer
+	// into the update channels.
+	netlinkBuffer = 16
+
+	excludedDevicePrefixes = []string{
+		defaults.HostDevice,
+		"cilium_",
+		"lo",
+		"lxc",
+		"cni",
+		"docker",
+		"sit",
+	}
+
+	// Route filter to look at all routing tables.
+	routeFilter = netlink.Route{
+		Table: unix.RT_TABLE_UNSPEC,
+	}
+	routeFilterMask = netlink.RT_FILTER_TABLE
+)
+
+type DevicesConfig struct {
+	Devices []string
+}
+
+type devicesControllerParams struct {
+	cell.In
+
+	Config      DevicesConfig
+	Log         logrus.FieldLogger
+	DB          statedb.DB
+	DeviceTable statedb.Table[*tables.Device]
+	RouteTable  statedb.Table[*tables.Route]
+
+	// NetNS is an optional network namespace handle. Used by tests to restrict the detection
+	// to specific namespace.
+	NetNS *netns.NsHandle `optional:"true"`
+
+	// netlinkFuncs is optional and used by tests to verify error handling behavior.
+	NetlinkFuncs *netlinkFuncs `optional:"true"`
+
+	// netlinkHandle is optional and used by tests to verify error handling behavior.
+	NetlinkHandle netlinkHandle `optional:"true"`
+}
+
+type devicesController struct {
+	params devicesControllerParams
+
+	log            logrus.FieldLogger
+	initialized    chan struct{}
+	filter         deviceFilter
+	l3DevSupported bool
+	ctx            context.Context
```

Comment:
Yep, this can as well be passed as parameter. The cancel function is needed here for the stop hook. I'll refactor this to use `pkg/hive/job` instead which will do the context management for us.